numbat/examples/inspect.rs

```diff
@@ -86,15 +79,126 @@ fn inspect_functions_in_module(ctx: &Context, module: String) {
+                .unwrap();
+
+            let extra_import = if !example_ctx
+                .resolver()
+                .imported_modules
+                .contains(&module_path)
+            {
+                format!("use {}\n", module)
+            } else {
+                "".into()
+            };
+            let _result = example_ctx
+                .interpret(&extra_import, CodeSource::Internal)
+                .unwrap();
+
+            if let Ok((statements, results)) =
+                example_ctx.interpret(&example_code, CodeSource::Internal)
+            {
+                //Format the example input
+                let example_input = format!(">>> {}", example_code);
+
+                //Encode the example url
+                let url_code = extra_import + &example_code;
+                let example_url = format!(
+                    "https://numbat.dev/?q={}",
+                    percent_encoding::utf8_percent_encode(
+                        &url_code,
+                        percent_encoding::NON_ALPHANUMERIC
+                    )
+                );
+
+                //Assemble the example output
+                let mut example_output = String::new();
+                example_output += "\n";
+
+                for statement in &statements {
+                    example_output += &plain_text_format(&statement.pretty_print(), true);
+                    example_output += "\n\n";
+                }
+
+                let result_markup = results.to_markup(
+                    statements.last(),
+                    &example_ctx.dimension_registry(),
+                    true,
+                    true,
+                );
+                example_output += &plain_text_format(&result_markup, false);
+
+                if results.is_value() {
+                    example_output += "\n";
+                }
+
+                //Print the example
+                if let Some(example_description) = example_description {
+                    println!(
+                        "* {}\n\n  <a href=\"{}\"><i class=\"fa fa-play\"></i> Run this example</a>",
+                        replace_equation_delimiters(example_description),
+                        example_url
+                    );
+                } else {
+                    println!(
+                        "* <a href=\"{}\"><i class=\"fa fa-play\"></i> Run this example</a>\n",
+                        example_url
+                    );
+                }
+
+                println!("  ```nbt");
+                for l in example_input.lines() {
+                    println!("    {}", l);
+                }
+                for l in example_output.lines() {
+                    println!("    {}", l);
+                }
+                println!("  ```");
+            } else {
+                eprintln!(
+                        "Warning: Example \"{example_code}\" of function {fn_name} did not run successfully."
+                    );
+            }
```

Comment:
I am not sure if this is the right choice here. Most often it is useful to check that examples run successfully, however in some cases one may want to deliberately trigger an error.

Reply:
I think we should fail hard here (instead of just showing a warning). Those examples are a great source of additional tests. Either we fail the generation of the documentation here (easy option), or we write a new unit test that checks all `@example`s (harder option, probably nicer for future development, since the feedback loop is quicker).

If we ever need failing examples (e.g. if we want to add an `@example` for the `error` function), we can think of a solution then. I don't see a strong need for this right now.

Reply:
Failing hard makes sense here I think, examples in the standard library should really be enforced to run without errors. I will change that.

I am not so sure how suitable the examples are as tests however. They do provide some basic check if the code runs without error but we can't assert that functions returns the right result so there is still a lot of room to make undetected mistakes.

> e.g. if we want to add an `@example` for the `error` function

Yes, this was the only real example that came to mind maybe next to hypothetical examples of how not to use a function so I would agree that there is not much need for it right now.